pkg/apiclients/testapi/testapi.go

```diff
@@ -0,0 +1,472 @@
+			errChan <- localErr
+			return
+		}
+		// h.finalTestID is already set if err is nil
+
+		result, err := h.fetchResultStatus(ctx, *finalTestID)
+		if err != nil {
+			localErr = fmt.Errorf("failed to fetch final test status for %s: %w", *finalTestID, err)
+			errChan <- localErr
+			return
+		}
+
+		h.setResult(result)
+		errChan <- nil
+	})
+
+	// Wait until the polling goroutine (started by runOnce.Do) finishes.
+	<-h.doneChan
+	return <-errChan
+}
+
+// Check status of a 200 OK response from GetJob.
+// Returns true if polling should stop due to the job itself erroring,
+// and an error if the job has errored. Otherwise, it returns false, nil.
+func handleJobInProgress(resp *GetJobResponse, jobID uuid.UUID) (stopPolling bool, err error) {
+	if resp.ApplicationvndApiJSON200 != nil {
+		status := resp.ApplicationvndApiJSON200.Data.Attributes.Status
+		if status == Errored {
+			return true, fmt.Errorf("job %s reported status 'errored'", jobID)
+		}
+	}
+	return false, nil
+}
+
+// Process a 303 See Other response from GetJob, extracting final Test ID from the Relationships.Test field.
+// Returns testID on success, or an error if testID is not found.
+func handleJobRedirect(resp *GetJobResponse, jobID uuid.UUID) (*uuid.UUID, error) {
+	if resp.ApplicationvndApiJSON303 != nil && resp.ApplicationvndApiJSON303.Data.Relationships != nil {
+		testID := resp.ApplicationvndApiJSON303.Data.Relationships.Test.Data.Id
+		if testID == uuid.Nil {
+			return nil, fmt.Errorf("job %s completed (303) but the final Test ID was missing in the response", jobID)
+		}
+		return &testID, nil
+	} else {
+		return nil, fmt.Errorf("job %s completed (303) but the final Test ID path was incomplete in the response", jobID)
+	}
+}
+
+// Create a detailed error message for unexpected HTTP statuses during polling.
+func formatUnexpectedJobStatusError(resp *GetJobResponse, jobID uuid.UUID) error {
+	errorBodyBytes := resp.Body
+	errMsg := fmt.Sprintf("unexpected status code %d polling job %s", resp.StatusCode(), jobID)
+	// Attempt to get more detail from structured error response
+	if resp.ApplicationvndApiJSON400 != nil && len(resp.ApplicationvndApiJSON400.Errors) > 0 {
+		errMsg = fmt.Sprintf("%s: %s", errMsg, resp.ApplicationvndApiJSON400.Errors[0].Detail)
+	} else if len(errorBodyBytes) > 0 {
+		// Fallback to raw body if no structured error
+		errMsg = fmt.Sprintf("%s, response: %s", errMsg, string(errorBodyBytes))
+	}
+	return fmt.Errorf("%s", errMsg)
+}
+
+// Query the job endpoint until we're redirected to its 'related' link containing results
+func (h *testHandle) pollJobToCompletion(ctx context.Context) (*uuid.UUID, error) {
+	ticker := time.NewTicker(h.client.config.PollInterval)
+	defer ticker.Stop()
+
+	getJobParams := &GetJobParams{Version: h.client.config.APIVersion}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context canceled while polling job %s: %w", h.jobID, ctx.Err())
```

Comment:
Suggestion: Introduce an error catalog error `Operation Cancelled`

Reply:
Can we pick that up in a followup? We've identified the need for new error categories here, let's add tickets to get them into the error catalog, and improve them here once they're agreed on there.

The error catalog is lacking a lot of definition for these kinds of general non-HTTP-service errors, and there will probably be a need for discussion to handle them thoughtfully there.

I don't think we should be blocking this work on that.

Reply:
Yes we can. Happy to split error handling from the PR as long as we follow up on it. It is an too important topic that gets easily forgotten.

@dotkas and I want to provide some guidelines to more consistent error message.